pkg/source/smardex/pool_simulator.go

```diff
@@ -0,0 +1,191 @@
+package smardex
+
+import (
+	"encoding/json"
+	"math/big"
+	"time"
+
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/entity"
+	poolpkg "github.com/KyberNetwork/kyberswap-dex-lib/pkg/source/pool"
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/util/bignumber"
+	"github.com/KyberNetwork/logger"
+)
+
+var now = time.Now
+
+func NewPoolSimulator(entityPool entity.Pool) (*PoolSimulator, error) {
+	var pair SmardexPair
+	if err := json.Unmarshal([]byte(entityPool.Extra), &pair); err != nil {
+		return nil, err
+	}
+
+	numTokens := len(entityPool.Tokens)
+	tokens := make([]string, numTokens)
+	reserves := make([]*big.Int, numTokens)
+	for i := 0; i < numTokens; i++ {
+		tokens[i] = entityPool.Tokens[i].Address
+		reserves[i] = bignumber.NewBig(entityPool.Reserves[i])
+	}
+
+	return &PoolSimulator{
+		Pool: poolpkg.Pool{
+			Info: poolpkg.PoolInfo{
+				Address:    entityPool.Address,
+				ReserveUsd: entityPool.ReserveUsd,
+				// SwapFee:    swapFee,
+				Exchange: entityPool.Exchange,
+				Type:     entityPool.Type,
+				Tokens:   tokens,
+				Reserves: reserves,
+				Checked:  false,
+			},
+		},
+		SmardexPair: SmardexPair{
+			PairFee:        pair.PairFee,
+			FictiveReserve: pair.FictiveReserve,
+			PriceAverage:   pair.PriceAverage,
+			FeeToAmount:    pair.FeeToAmount,
+		},
+		gas: DefaultGas,
+	}, nil
+}
+
+func (p *PoolSimulator) CalcAmountOut(tokenAmountIn poolpkg.TokenAmount, tokenOut string) (*poolpkg.CalcAmountOutResult, error) {
+	if tokenAmountIn.Token == tokenOut {
+		return nil, ErrSameAddress
+	}
+
+	if isZero(tokenAmountIn.Amount) {
+		return nil, ErrZeroAmount
+	}
+
+	var zeroForOne bool
+	if tokenAmountIn.Token == p.GetTokens()[0] {
+		zeroForOne = true
+	}
+
+	var (
+		fictiveReserveIn   *big.Int = p.FictiveReserve.FictiveReserve0
+		fictiveReserveOut  *big.Int = p.FictiveReserve.FictiveReserve1
+		priceAverageIn     *big.Int = p.PriceAverage.PriceAverage0
+		priceAverageOut    *big.Int = p.PriceAverage.PriceAverage1
+		balanceIn          *big.Int = new(big.Int).Sub(p.GetReserves()[0], p.FeeToAmount.Fees0)
+		balanceOut         *big.Int = new(big.Int).Sub(p.GetReserves()[1], p.FeeToAmount.Fees1)
```

Comment:
Since the `getReserves()` function already calculate:
```
        reserve0_ = IERC20(token0).balanceOf(address(this)) - feeToAmount0;
        reserve1_ = IERC20(token1).balanceOf(address(this)) - feeToAmount1;
```

I think these lines should be changed to:
```suggestion
		balanceIn          *big.Int = p.GetReserves()[0]
		balanceOut         *big.Int = p.GetReserves()[1]
```

Reply:
Fixed, thank you.